README.md
# A simple flash-based VGA display

![A picture of a cat's head, surrounded by a kind of sunburst pattern,
being displayed on a montior](./demo.jpg)

The idea of this project is to generate a basic VGA-like signal that a
monitor will display by putting the signal in a flash memory, and then
using counters to run through the address space. Simple, eh?

## VGA signal timing

We have h-sync and v-sync signals, with their respective front and
back porches. The main trick is the find a resolution that fits nicely
into a flash memory, so that the flash memory is a multiple of the
screen size - not the visible screen size, but the screen size with
all the sync bits.

A bit of twiddling gives a 384x240 visible display that is 512x256
once all the sync etc. is factored in.

## Clocking

A 32MHz clock is used as a dot clock for a 768x480 image at 60 FPS
(1024x512 when sync signals are factored in). The Flash cannot handle
a 32MHz data rate, but we clock it every other cycle and do scan-line
doubling to display a 384x240 image.

## VGA levels

H- and V-sync signals are straightforward to produce, as they're
apparently supposed to be TTL. That leaves 6 bits for 3 channels of
colour, or 2 bits per channel. A pair of resistors on each channel
makes the world's simplest DAC, with values chosen to give a 0.7V peak
when terminated by a 75 ohm resistor when both bits are set.

## Test pattern

gen_test.lua generates a simple RGB "checkerboard" displaying the
range of available colours.

gen_rom.lua generates a ROM image given a Gimp .data file (I couldn't
find a nice low-dep Lua image library). It simply takes the top two
bits of each pixel value to generate the colour.

prep_img.lua converts a .data image into another .data image that uses
the 6-bit colour space that the circuit can display. It uses dithering
to improve the generated image. Generating a ROM image from an image
that has been prepared with 'prep_img' should produce a rather better
result.

## Circuit design

The circuit breadboarded so far has three chained 74HC counters
supplying the address to a 29040 flash chip. A 74HC574 latches the
values for the display. (A 74HC574 triggers on the wrong edge to avoid
transients, but it's what I had to hand, and seems to work fine.)

## PCB

![A 3D render of the PCB](./hw/images/3d.png)

I sent the files in hw/gerbers to http://smart-prototyping.com/, and
they quickly sent me back PCBs which did the job. There are PDFs of
the Gerbers in hw/images.

### Assembly notes

The PCB is pretty straightforward to assemble (no surface mount!) as
long as you pay attention to the orientations, as I oriented things to
make routing easy, rather than assembly.

There should be enough space for a ZIF for the flash chip, although I
only used an IC socket.

The PCB has been designed to support 8-pin-sized oscillators as well
as 14-pin-sized oscillators. However, I installed the larger size.
Take a careful look at the PCB to see how to install the smaller
crystal if that's what you have. Check that power and ground reach the
appropriate pins.

Note that pin 1 of the power connector is ground, and pin 3 is Vcc.
This is something I wasn't paying attention to when setting up the
schematic (nor did I label it on the silkscreen).

### Animation speed

There are two headers on the board that you can jumper to set the
animation speed. They look like this:

```
 . . .   . . .
   P3      P2
```

The possibilities are:

```
 === .   === . Slow

 === .   . === Medium

 . ===   . === Fast
```

Note that only "medium" displays the 4 frames in the order they are in
memory (i.e. 1 2 3 4 1 2 3 4 ...). In the other two modes, the order
is 1 3 2 4 1 3 2 4 ... (if you're wondering why, it's because the
2-bit counter is bit-reversed). You will need to take this into
account when programming your animations.

## ROM image generation

The 'build.sh' script will generate a nice spinny [Bad Advice
cat](http://knowyourmeme.com/memes/bad-advice-cat) ROM image. If
you're using medium-speed mode, you'll need to pass the 'swap' flag.

I was originally hoping to script up the image wrangling mostly in
Gimp, but it turns out the script-fu support is rubbish. You can't
even [load a raw .data file in a non-interactive
script](https://github.com/marcelteun/GIMP-raw-file-load/blob/5f02e7607b1645d54b11bb9f01d44e362abca2d1/file-raw-load.c#L237).
So, it's just raw data files.

cat.data.gz was generated from cat.png using the Gimp interactively.
Which was me just scrubbing the background off the Bad Advice Cat
image. Minor copyright infringement is probably involved.

If you want to create your own image, I'm sure you can work out how to
do it by looking at build.sh and the lua scripts.


hw/CHANGELOG.md
# v0

Very first circuit I created. Low frequency (2MHz), has a few problems
(black stripes between the wide "pixels").

Note that the R and B channels are swapped around compared to what I
actually built (and thus what I wrote scripts for), as it's equivalent
and makes the schematic a lot easier to draw.

# v0.1

I added a 74574 latch to the output, so that the transients from the
output of the Flash aren't sent over the VGA lines, which resulted in
black stripes on the screen.

74574 = no stripes. Hurrah!

# v0.2

Switched from 2MHz oscillator to 16MHz oscillator, making the pixels
8x narrower. A 32MHz oscillator would give square pixels, but the
29040 takes 55ns to read, so that would be too fast.

Oscilloscope output looked good, but video didn't display. The
oscillator's output didn't look like a great square wave, but putting
an inverter or two on the 74754 latch's clock line made the display
work. It's not totally clear to me if this was because it helped
square up the wave, or perhaps because it fixed up a clock phase issue
- I need to investigate more.

By attaching the high address line to the highest bit of the counter,
we can create a simple animation with 2 frames.

# v0.3

Moved the inverter to directly after the oscillator. The display
works. This suggests that it's the signal shape, not phase issues
between different chips, that causes the problem. It seems only the
latch is annoyed by the shape of the oscillator output being
insufficiently square.

# v0.4

This revision is an optimisation, removing the inverter. As the
problem was feeding the oscillator output directly into the latch, I
removed the oscillator and did it another way: Replace the 16MHz
oscillator with a 32MHz oscillator, move all the address pins along to
get the same frequency counting, and use the lowest bit of the counter
as the clock for the latch. It seems to do the trick.

A downside is that the animation now only runs at twice the speed as
before (a bit manic), but it's a trade-off I'm willing to take to
eliminate a whole IC from a small design.

# v0.5

Rearranged the counter to perform scan-line doubling, which gives us
space for 4 frames in the Flash memory, as well as squarer pixels.

# v0.6

Added jumpers to allow control of the animation rate.

# v0.7

Swapped data lines around to simplify PCB routing.

# v0.8

Update to newer version of Kicad.


src/gen_test.lua
-- Generate an EEPROM image suitable for display on VGA

local h_front_porch = 16
local h_sync = 64
local h_back_porch = 48
local h_visible = 384

local v_front_porch = 1
local v_sync =  1
local v_back_porch = 14
local v_visible = 240

local rom_size = 512 * 1024

local out_file = "vga.bin"

local h_total = h_front_porch + h_sync + h_back_porch + h_visible
print ("Total horizontal resolution: " .. h_total)

local v_total = v_front_porch + v_sync + v_back_porch + v_visible
print ("Total vertical resolution: " .. v_total)

local total = h_total * v_total
print ("Total data: " .. total)

if rom_size % total ~= 0 then
  print ("ROM size (" .. rom_size .. ")is not a multiple of image size (" ..
    total .. ")")
  os.exit(1)
end

local repeats = rom_size / total

local v_sync_val = 0x80
local h_sync_val = 0x40

function pixel_at(x, y)
  -- x value controls bottom two bits
  local low = math.floor((x - 1) * 4 / h_visible)
  -- y value controls next 4 bits
  local high = math.floor((y - 1) * 16 / v_visible)
  return high * 4 + low
end

function write_byte(out, b)
  out:write(string.char(b))
end

function write_line(out, mode, line_num)
  local base_val = mode == "sync" and v_sync_val or 0x00

  for i = 1, h_sync do
    write_byte(out, base_val + h_sync_val)
  end
  for i = 1, h_back_porch do
    write_byte(out, base_val)
  end
  for i = 1, h_visible do
    local val = mode == "visible" and pixel_at(i, line_num) or 0
    write_byte(out, base_val + val)
  end
  for i = 1, h_front_porch do
    write_byte(out, base_val)
  end
end

function write_frame(out)
  for i = 1, v_sync do
    write_line(out, "sync", i)
  end
  for i = 1, v_back_porch do
    write_line(out, "blank", i)
  end
  for i = 1, v_visible do
    write_line(out, "visible", i)
  end
  for i = 1, v_front_porch do
    write_line(out, "blank", i)
  end
end

function main(filename)
  local out = assert(io.open(out_file, "wb"))

  for r = 1, repeats do
    write_frame(out)
  end

  assert(out:close())
end

main(out_file)


tools/gen_test.lua
-- Generate an EEPROM image suitable for display on VGA
--
-- NB: Pattern produced differs from photos I've taken because the bit
-- ordering has changed since the original prototype.

local h_front_porch = 16
local h_sync = 64
local h_back_porch = 48
local h_visible = 384

local v_front_porch = 1
local v_sync =  1
local v_back_porch = 14
local v_visible = 240

local rom_size = 512 * 1024

local out_file = "vga.bin"

local h_total = h_front_porch + h_sync + h_back_porch + h_visible
print ("Total horizontal resolution: " .. h_total)

local v_total = v_front_porch + v_sync + v_back_porch + v_visible
print ("Total vertical resolution: " .. v_total)

local total = h_total * v_total
print ("Total data: " .. total)

if rom_size % total ~= 0 then
  print ("ROM size (" .. rom_size .. ")is not a multiple of image size (" ..
    total .. ")")
  os.exit(1)
end

local repeats = rom_size / total

local v_sync_val = 0x02
local h_sync_val = 0x01

function pixel_at(x, y)
  -- x value controls bottom two bits
  local low = math.floor((x - 1) * 4 / h_visible)
  -- y value controls next 4 bits
  local high = math.floor((y - 1) * 16 / v_visible)
  return (high * 4 + low) * 4
end

function write_byte(out, b)
  out:write(string.char(b))
end

function write_line(out, mode, line_num)
  local base_val = mode == "sync" and v_sync_val or 0x00

  for i = 1, h_sync do
    write_byte(out, base_val + h_sync_val)
  end
  for i = 1, h_back_porch do
    write_byte(out, base_val)
  end
  for i = 1, h_visible do
    local val = mode == "visible" and pixel_at(i, line_num) or 0
    write_byte(out, base_val + val)
  end
  for i = 1, h_front_porch do
    write_byte(out, base_val)
  end
end

function write_frame(out)
  for i = 1, v_sync do
    write_line(out, "sync", i)
  end
  for i = 1, v_back_porch do
    write_line(out, "blank", i)
  end
  for i = 1, v_visible do
    write_line(out, "visible", i)
  end
  for i = 1, v_front_porch do
    write_line(out, "blank", i)
  end
end

function main(filename)
  local out = assert(io.open(out_file, "wb"))

  for r = 1, repeats do
    write_frame(out)
  end

  assert(out:close())
end

main(out_file)


build.sh
#!/bin/sh
#
# Script to generate the ROM contents needed to display the image on
# the hardware.
#

set -e

mkdir -p out
cd out

# Generate the "sunburst"
lua ../tools/gen_img.lua

# Swap ordering of frames if you're running in "medium" speed
# animation mode.
if [[ "$1" == "swap" ]]
then
  mv out2.data saved.data
  mv out3.data out2.data
  mv saved.data out3.data
fi

# Extract the compressed cat image data
gunzip -c ../data/cat.data.gz > cat.data

# Overlay the cat
for IMAGE in out?.data
do
  lua ../tools/compose.lua -1 cat.data -2 $IMAGE -o overlay_$IMAGE
done

# Dither the source files for the limited bit depth we can display
for IMAGE in overlay_out?.data
do
  lua ../tools/prep_img.lua -i $IMAGE -o dithered_$IMAGE
done

# Convert from raw files to raster images that can be fed as a VGA
# signal.
for IMAGE in dithered_overlay_out?.data
do
  lua ../tools/gen_rom.lua -i $IMAGE -o $IMAGE.bin
done

# Place the images back to back to create the final ROM
cat dithered_overlay_out?.data.bin > spin.bin


src/compose.lua
-- Compose a 4-channel RGBA raw image over a 3-channel RGB image,
-- generating a new 3-channel RGB image.

local in1_file
local in2_file
local out_file

do
    local i = 1
    while i <= #arg do
        if arg[i] == "-1" then
            i = i + 1
            in1_file = arg[i]
        elseif arg[i] == "-2" then
            i = i + 1
            in2_file = arg[i]
        elseif arg[i] == "-o" then
            i = i + 1
            out_file = arg[i]
        else
            print(arg[0] .. ": Unrecognised option: " .. arg[i])
            os.exit(1)
        end
        i = i + 1
    end
end

if in1_file == nil then
    print(arg[0] .. ": Overlay input filename required")
    os.exit(1)
end

if in2_file == nil then
    print(arg[0] .. ": Base input filename required")
    os.exit(1)
end

if out_file == nil then
    print(arg[0] .. ": Output filename required")
    os.exit(1)
end

------------------------------------------------------------------------
-- Do the work

function read_pixel_rgba(fin)
  local pixel = fin:read(4)
  if pixel == nil then
    return nil
  end
  r = pixel:byte(1)
  g = pixel:byte(2)
  b = pixel:byte(3)
  a = pixel:byte(4)
  return { r = r, g = g, b = b, a = a }
end

function read_pixel_rgb(fin)
  local pixel = fin:read(3)
  if pixel == nil then
    return nil
  end
  r = pixel:byte(1)
  g = pixel:byte(2)
  b = pixel:byte(3)
  return { r = r, g = g, b = b }
end

function write_pixel_rgb(fout, p)
  local data = string.char((p.r + 0.5), (p.g + 0.5), (p.b + 0.5))
  fout:write(data)
end

-- Pixel maths!

function overlay(p1, p2)
  local w1 = p1.a / 255
  local w2 = 1 - w1
  return { r = p1.r * w1 + p2.r * w2,
           g = p1.g * w1 + p2.g * w2,
           b = p1.b * w1 + p2.b * w2 }
end

function main(in_file, out_file)
  local fin1 = assert(io.open(in1_file, "rb"))
  local fin2 = assert(io.open(in2_file, "rb"))
  local fout = assert(io.open(out_file, "wb"))

  while true do
    local front = read_pixel_rgba(fin1)
    local base = read_pixel_rgb(fin2)
    if front == nil and base == nil then
      break
    end
    assert(front ~= nil and base ~= nil)
    write_pixel_rgb(fout, overlay(front, base))
  end

  assert(fin1:close())
  assert(fin2:close())
  assert(fout:close())
end

main(in_file, out_file)


src/gen_img.lua
-- Generates an animated spinning radial pattern useful for
-- psychodelic animation.

local width = 384
local height = 240

local c1 = { r = 244, g = 97, b = 27 }
local c2 = { r = 255, g = 235, b = 1 }
local num_rays = 15
local num_phases = 4

local out_file = "out%d.data"

function write_pixel(fout, p)
  local data = string.char((p.r), (p.g), (p.b))
  fout:write(data)
end

-- Interpolate between from and to, based on angle.
function wobble(angle, from, to)
  value = math.cos(angle * num_rays)
  -- Bunch the values up around the extremes by sqrting the abs value.
  if value < 0 then
    value = -math.sqrt(math.sqrt(-value))
  else
    value =  math.sqrt(math.sqrt(value))
  end
  -- Normalise 0..1
  value = value / 2.0 + 0.5
  -- And linear interpolate
  return from * (1 - value) + to * value
end

function main(out_file)
  for i = 1, num_phases do
    local file_name = out_file:format(i)
    local fout = assert(io.open(file_name, "wb"))

    for y = 1, height do
      for x = 1, width do
        angle = math.atan2(y - height / 2, x - width / 2)
        -- Fix NaN
        if angle ~= angle then angle = 0.0 end
        -- Add phase for animation
        angle = angle + 2 * math.pi * i / num_phases
        -- and colour pixel...
        pixel = {
          r = wobble(angle, 244, 255),
          g = wobble(angle, 97, 235),
          b = wobble(angle, 27, 1)
        }
        write_pixel(fout, pixel)
      end
    end

    assert(fout:close())
  end
end

main(out_file)


tools/prep_img.lua
-- Convert an image into the form that we'll put on the Flash.

-- Basically, uses Floyd-Steinberg to reduce the colour to 6 bits, to
-- get an image we can display.
--
-- As input, it takes a file saved in Gimp's raw .data format (as I
-- couldn't find a low-dep image processing library for Lua), and
-- produces a file in the same format.

local width = 384
local height = 240
local colour_levels = 4
local in_file
local out_file

do
    local i = 1
    while i <= #arg do
        if arg[i] == "-i" then
            i = i + 1
            in_file = arg[i]
        elseif arg[i] == "-o" then
            i = i + 1
            out_file = arg[i]
        elseif arg[i] == "-w" then
            i = i + 1
            width = arg[i]
        elseif arg[i] == "-h" then
            i = i + 1
            height = arg[i]
        elseif arg[i] == "-c" then
            i = i + 1
            colour_levels = arg[i]
        else
            print(arg[0] .. ": Unrecognised option: " .. arg[i])
            os.exit(1)
        end
        i = i + 1
    end
end

if in_file == nil then
    print(arg[0] .. ": Input filename required")
    os.exit(1)
end

if out_file == nil then
    print(arg[0] .. ": Output filename required")
    os.exit(1)
end

------------------------------------------------------------------------
-- Do the work

function read_pixel(fin)
  local pixel = fin:read(3)
  r = pixel:byte(1)
  g = pixel:byte(2)
  b = pixel:byte(3)
  return { r = r, g = g, b = b }
end

function write_pixel(fout, p)
  local data = string.char((p.r), (p.g), (p.b))
  fout:write(data)
end

-- Quantize one component
function adjust(x)
  local val = math.floor((x / 255) * colour_levels + 0.5)
  local target = math.floor((val / colour_levels) * 255 + 0.5)
  return target
end

-- Quantize a pixel
function quantize(p)
  return { r = adjust(p.r), g = adjust(p.g), b = adjust(p.b) }
end

-- Pixel maths!
local black = { r = 0.0, g = 0.0, b = 0.0 }

function add(p1, p2)
  return { r = p1.r + p2.r, g = p1.g + p2.g, b = p1.b + p2.b }
end

function mult(x, p)
  return { r = x * p.r, g = x * p.g, b = x * p.b }
end

function main(in_file, out_file)
  local fin = assert(io.open(in_file, "rb"))
  local fout = assert(io.open(out_file, "wb"))

  -- Initialise error diffuser thing
  local prev_line = {}
  for x = 1, width do
    prev_line[x] = black
  end

  for y = 1, height do
    local prev_pixel = black
    for x = 1, width do
      local pixel = read_pixel(fin)
      pixel = add(add(pixel, prev_line[x]), prev_pixel)
      local qpixel = quantize(pixel)
      local err = add(pixel, mult(-1, qpixel))
      local prev_pixel = mult(0.5, err)
      prev_line[x] = mult(0.5, err)
      write_pixel(fout, qpixel)
    end
  end

  assert(fin:close())
  assert(fout:close())
end

main(in_file, out_file)


src/gen_rom.lua
-- Generate a ROM file from a .data file.
--
-- Processes a Gimp raw .data file into a ROM image. Assumes all the
-- dithering etc. has been done already (e.g. by prep_img.lua).
--
-- Note that this only generates one frame, but you can concatenate
-- two frames to fill a 512KB device.

local h_front_porch = 16
local h_sync = 64
local h_back_porch = 48
local h_visible = 384

local v_front_porch = 1
local v_sync =  1
local v_back_porch = 14
local v_visible = 240

------------------------------------------------------------------------
-- Argument reading

local in_file
local out_file

do
    local i = 1
    while i <= #arg do
        if arg[i] == "-i" then
            i = i + 1
            in_file = arg[i]
        elseif arg[i] == "-o" then
            i = i + 1
            out_file = arg[i]
        elseif arg[i] == "-w" then
            i = i + 1
            width = arg[i]
        elseif arg[i] == "-h" then
            i = i + 1
            height = arg[i]
        elseif arg[i] == "-c" then
            i = i + 1
            colour_levels = arg[i]
        else
            print(arg[0] .. ": Unrecognised option: " .. arg[i])
            os.exit(1)
        end
        i = i + 1
    end
end

if in_file == nil then
    print(arg[0] .. ": Input filename required")
    os.exit(1)
end

if out_file == nil then
    print(arg[0] .. ": Output filename required")
    os.exit(1)
end

------------------------------------------------------------------------
-- Image reading part
--

local data = {}

function read_pixel(f, x, y)
  local pixel = f:read(3)
  r = pixel:byte(1)
  g = pixel:byte(2)
  b = pixel:byte(3)
  r = math.floor(r + 0.5)
  g = math.floor(g + 0.5)
  b = math.floor(b + 0.5)
  -- Extract top two bits.
  r = math.floor(r / 64) % 4
  g = math.floor(g / 64) % 4
  b = math.floor(b / 64) % 4
  -- Assemble and store the value
  data[y][x] = (r * 16 + g * 4 + b) * 4
end

function read_image(f)
  for y = 1, v_visible do
    data[y] = {}
    for x = 1, h_visible do
      read_pixel(f, x, y)
    end
  end
end

------------------------------------------------------------------------
-- ROM writing part
--

local v_sync_val = 0x02
local h_sync_val = 0x01

function pixel_at(x, y)
  return data[y][x]
end

function write_byte(out, b)
  out:write(string.char(b))
end

function write_line(out, mode, line_num)
  local base_val = mode == "sync" and v_sync_val or 0x00

  for i = 1, h_sync do
    write_byte(out, base_val + h_sync_val)
  end
  for i = 1, h_back_porch do
    write_byte(out, base_val)
  end
  for i = 1, h_visible do
    local val = mode == "visible" and pixel_at(i, line_num) or 0
    write_byte(out, base_val + val)
  end
  for i = 1, h_front_porch do
    write_byte(out, base_val)
  end
end

function write_frame(out)
  for i = 1, v_sync do
    write_line(out, "sync", i)
  end
  for i = 1, v_back_porch do
    write_line(out, "blank", i)
  end
  for i = 1, v_visible do
    write_line(out, "visible", i)
  end
  for i = 1, v_front_porch do
    write_line(out, "blank", i)
  end
end

------------------------------------------------------------------------
-- Put it all together...
--

function main(filename)
  local fin = assert(io.open(in_file, "rb"))
  read_image(fin)
  assert(fin:close())

  local fout = assert(io.open(out_file, "wb"))
  write_frame(fout)
  assert(fout:close())
end

main(in_file, out_file)
